File: 17_dynamic_programming/min_abs_sum.py
def solution(a):
    n = len(a)
    if n == 0:
        return 0
    a = map(abs, a)
    cnts = [0] * (101)
    for i in a:
        cnts[i] += 1
    s = sum(a)
    ub = s // 2
    dm = [-1] * (ub + 1)
    dm[0] = 0
    for i in range(1, 101):
        if cnts[i] < 1:
            continue
        for j in range(ub + 1):
            if dm[j] >= 0:
                dm[j] = cnts[i]
            elif (j >= i and dm[j-i] > 0):
                dm[j] = dm[j-i] - 1
    for i in range(ub, -1, -1):
        if dm[i] >= 0:
            return s - 2 * i

File: 17_dynamic_programming/min_abs_sum.cpp
#include <vector>

using namespace std;

int abs_sum(vector<int> &a) {
  int rt = 0;
  for (auto i : a) {
    rt += abs(i);
  }
  return rt;
}

int solution(vector<int> &a) {
  int n = a.size();
  int s = abs_sum(a);
  int m = s / 2;
  vector<int> dp(m + 1, -1);
  vector<int> cnts(101, 0);
  for (int i = 0; i < n; i++) {
    cnts[abs(a[i])]++;
  }
  dp[0] = 0;
  for (int i = 1; i < 101; i++) {
    if (cnts[i] == 0) {
      continue;
    }
    for (int j = 0; j < m + 1; j++) {
      if (dp[j] >= 0) {
        dp[j] = cnts[i];
      } else if (j >= i && dp[j - i] > 0){
        dp[j] = dp[j - i] - 1;
      }
    }
  }
  for (int i = m; i >= 0; i--) {
    if (dp[i] >= 0) {
      return s - 2 * i;
    }
  }
}
